server/src/main/java/org/opensearch/action/search/SearchRequestListenerManager.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.search;
+
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Setting;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+
+/**
+ * SearchRequestListenerManager manages listeners registered to search requests,
+ * and is responsible for creating the {@link SearchRequestOperationsListener.CompositeListener}
+ * with the all listeners enabled at cluster-level and request-level.
+ *
+ *
+ * @opensearch.internal
+ */
+public class SearchRequestListenerManager {
+
+    private final ClusterService clusterService;
+    public static final String SEARCH_PHASE_TOOK_ENABLED_KEY = "search.phase_took_enabled";
+    public static final Setting<Boolean> SEARCH_PHASE_TOOK_ENABLED = Setting.boolSetting(
+        SEARCH_PHASE_TOOK_ENABLED_KEY,
+        false,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    private final List<SearchRequestOperationsListener> searchRequestListenersList;
+
+    public SearchRequestListenerManager(
+        ClusterService clusterService
+    ) {
+        this.clusterService = clusterService;
+        searchRequestListenersList = new ArrayList<>();
+    }
+
+    /**
+     * Add multiple {@link SearchRequestOperationsListener} to the searchRequestListenersList.
+     * Those enabled listeners will be executed during each search request.
+     *
+     * @param listeners Multiple SearchRequestOperationsListener object to add.
+     * @throws IllegalArgumentException if any input listener is null or already exists in the list.
+     */
+    public void addListeners(SearchRequestOperationsListener... listeners) {
+        for (SearchRequestOperationsListener listener : listeners) {
+            if (listener == null) {
+                throw new IllegalArgumentException("listener must not be null");
+            }
+            if (searchRequestListenersList.contains(listener)) {
+                throw new IllegalArgumentException("listener already added");
```

Comment:
The order of listeners could sometimes matter, thus I was hesitated to use `set` instead of `list`. Also, the number of listeners got added to the manager won't be a very large number, the time complexity of using list should be fine.

Reply:
Do you need this check at all? `SearchRequestOperationsListener` doesn't force implementations to override equals and hash code, so this is falling back to identity-based equality which means it probably isn't adding a lot of value. What are you trying to guard against here?

Reply:
@andrross make sense to me, let me remove the check.